worker/docker_image_manager.py

```diff
@@ -0,0 +1,140 @@
+        }
+        with open(self._state_file, 'w') as f:
+            json.dump(state, f)
+
+    def touch_image(self, image_id):
+        """
+        Update the last-used date of an image to be the current date.
+        """
+        # Writes must be exclusive with other reads and writes
+        with self._lock:
+            now = time.time()
+            self._last_used[image_id] = now
+            self._save_state()
+        logger.debug('touched image %s at %f' % (image_id, now))
+
+    def start_cleanup_thread(self):
+        self._stop_cleanup = False
+        self._cleanup_thread = threading.Thread(target=DockerImageManager._do_cleanup, args=[self])
+        self._cleanup_thread.start()
+
+    def stop_cleanup_thread(self):
+        if self._cleanup_thread is None:
+            return
+        with self._lock:
+            self._stop_cleanup = True
+        self._cleanup_thread.join()
+        self._cleanup_thread = None
+
+    def _should_stop_cleanup(self):
+        with self._lock:
+            return self._stop_cleanup
+
+    def _get_free_bytes(self):
+        """
+        Return bytes free on disk containing state file.
+        """
+        fd = os.open(self._docker_root_dir, os.O_RDONLY)
+        stat = os.fstatvfs(fd)
+        os.close(fd)
+        return stat.f_bavail * stat.f_frsize
+
+    def _do_cleanup(self):
+        """
+        Periodically clean up the oldest images when the disk is close to full.
+        """
+        logging.debug('Image cleanup thread started')
+        while not self._should_stop_cleanup():
+            # Start disk usage reduction loop
+            # Iteratively try to remove the oldest images until the disk usage
+            # is within the limit.
+            try:
+                while True:
+                    free_bytes = self._get_free_bytes()
+                    logging.debug('disk free: %s (min %s)' %
+                                  (size_str(free_bytes), size_str(self._min_free_bytes)))
+
+                    if len(self._last_used) > 0 and free_bytes < self._min_free_bytes:
+                        with self._lock:
+                            # Pick the stalest image
+                            image_id = min(self._last_used, key=lambda id_: self._last_used[id_])
+
+                            # Remove that image
+                            self._docker.remove_image(image_id)
+                            del self._last_used[image_id]
+                            self._save_state()
+                            logging.info('removed image %s' % image_id)
+                    else:
+                        # Break out of the loop when disk usage is normal.
+                        break
+            except Exception:
+                # Print the error then go to sleep and try again next time.
+                traceback.print_exc()
+
+            # Sleep for 10 seconds, allowing interruptions every second.
+            for _ in xrange(0, self._cleanup_sleep_seconds):
```

Comment:
Do we need this logic, as opposed to just sleeping 1 second in the main loop?  All we have to do is check free space, which should be fast.